### ISSUE TITLE
_Py_DECREF_SPECIALIZED doesn't respect pymalloc tracking

### ISSUE THREAD

**Issue by pablogsal** [python/cpython#125703]
PR https://github.com/python/cpython/pull/30872 added `_Py_DECREF_SPECIALIZED` which calls `destruct` over the object. When `destruct` is `PyObject_Free` this skips `tracemalloc` counting reporting that the memory is alive. This also makes debuggers segfault because they think that the object is alive because they did not get a notification. This is the code that's not getting called:

https://github.com/python/cpython/blob/2e950e341930ea79549137d4d3771d5edb940e65/Objects/object.c#L2924-L2928

This may also qualify a regression as makes tracemalloc detect incorrect memory usage for these objects

### Linked PRs
* gh-125704
* gh-125705
* gh-125706
* gh-125707
* gh-125712
* gh-125791

**Comment by pablogsal** [comment]
CC: @sweeneyde @Yhg1s

**Comment by pablogsal** [comment]
CC: @mdboom @markshannon

**Comment by pablogsal** [comment]
I suspect that we are missing more calls to tracemalloc hooks when the memory is reused. Basically every call to `_Py_ForgetReference` needs to have a call to tracemalloc hooks.

**Comment by pablogsal** [comment]
3.11 and 3.12 are not affected because there is no `tracemalloc_remove_trace`

**Comment by sweeneyde** [comment]
The macro version in ceval.c may also be affected:

https://github.com/python/cpython/blob/f8ba9fb2ce6690d2dd05b356583e8e4790badad7/Python/ceval.c#L93

I might be misunderstanding how the macros expand, but I believe this macro is unused anyway.

**Comment by markshannon** [comment]
@pablogsal You seem to be claiming that https://github.com/python/cpython/pull/30872 caused a regression in a feature added in https://github.com/python/cpython/pull/115945.
How can an earlier change cause a regression in a feature added later?

Combined with such comments as [this](https://github.com/python/cpython/pull/125704#issuecomment-2423011866), it feels like you are trying to shift the blame for this issue.

**Comment by pablogsal** [comment]
> @pablogsal You seem to be claiming that https://github.com/python/cpython/pull/30872 caused a regression in a feature added in https://github.com/python/cpython/pull/115945.
> 
> How can an earlier change cause a regression in a feature added later?
> 
> 
> 
> Combined with such comments as [this](https://github.com/python/cpython/pull/125704#issuecomment-2423011866), it feels like you are trying to shift the blame for this issue.
> 
> 

I am not shifting blame for anything. Let me explain yet again the problem:

https://github.com/python/cpython/pull/30872 caused a regression for tracemalloc AND therefore also affect the hooks I added https://github.com/python/cpython/pull/115945. 

The regression of the feature is for tracemalloc which was added many releases ago.

**Comment by vstinner** [comment]
I think that we should first make the code correct (fix tracemalloc/the callback), then make the code fast.

**Comment by markshannon** [comment]
@vstinner I don't see any value in making trite comments like "first make the code correct, then make the code fast"?

It suggests that correctness and performance are totally unrelated. Often considering performance will lead to a more streamlined implementation that can be more maintainable.

For example, if tracemalloc was entirely built on PEP 445 (as PEP 454 states it would), there would be no issue in the first place as all allocations and frees would be done through the PEP 445 function pointer(s).

**Comment by pablogsal** [comment]
Ok, so @markshannon and I had a catch up call about this. We discussed our views and we clarified our positions a bit better. Here is what we have agreed:

* Mark is concerned about how these hooks can impact performance, specially in the future once we have a JIT. One of the thinks he is concerned is that if we increase the size of `_Py_DECREF_SPECIALIZED` then the JIT will have a lot of repeated code to optimise and that will be bad.
* Mark thinks there is a better way to place these hooks that will have less performance impact. I will open a new issue where we can track these improvements for 3.14.
* We have agreed that for 3.13 as the JIT is experimental and pyperformance didn't show any regression above noise (noise at least defined with ~0.5% change) for the [fix](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20241018-3.14.0a1%2B-f8ba9fb/bm-20241018-linux-x86_64-python-f8ba9fb2ce6690d2dd05-3.14.0a1%2B-f8ba9fb-vs-base.svg) or the original PR then we will go with the combo of https://github.com/python/cpython/pull/125704 and https://github.com/python/cpython/pull/125712 since this fixes the issue without introducing a lot of trouble.
